core/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/entity/decorator/FilteringEntityByEntitySelector.java

```diff
@@ -0,0 +1,376 @@
+                countIterations, countAttempts, countAttempts / (double) countIterations, countSuccess,
+                countSuccess / (double) countIterations, countFails,
+                countFails / (double) countIterations);
+    }
+
+    // ************************************************************************
+    // Worker methods
+    // ************************************************************************
+
+    @Override
+    public EntityDescriptor<Solution_> getEntityDescriptor() {
+        return childEntitySelector.getEntityDescriptor();
+    }
+
+    @Override
+    public long getSize() {
+        return allEntities.size();
+    }
+
+    @Override
+    public boolean isCountable() {
+        return childEntitySelector.isCountable();
+    }
+
+    @Override
+    public boolean isNeverEnding() {
+        return childEntitySelector.isNeverEnding();
+    }
+
+    /**
+     * The expected replayed entity corresponds to the selected entity when the replaying selector has the next value.
+     * Once it is selected, it will be reused until a new entity is replayed by the recorder selector.
+     */
+    private Object selectReplayedEntity() {
+        var iterator = replayingEntitySelector.iterator();
+        if (iterator.hasNext()) {
+            this.replayedEntity = iterator.next();
+        }
+        return replayedEntity;
+    }
+
+    @Override
+    public Iterator<Object> endingIterator() {
+        return new OriginalFilteringValueRangeIterator<>(this::selectReplayedEntity, childEntitySelector.iterator(),
+                basicVariableDescriptors, valueRangeManager);
+    }
+
+    @Override
+    public Iterator<Object> iterator() {
+        if (randomSelection) {
+            if (!childEntitySelector.isNeverEnding()) {
+                throw new IllegalArgumentException(
+                        "Impossible state: childEntitySelector must provide a never-ending approach.");
+            }
+            // Experiments have shown that a large number of attempts do not scale well,
+            // and 10 seems like an appropriate limit. 
+            // So we won't spend excessive time trying to generate a single move for the current selection.
+            // If we are unable to generate a move, the move iterator can still be used in later iterations.
+            return new RandomFilteringValueRangeIterator<>(this::selectReplayedEntity, childEntitySelector.iterator(),
+                    basicVariableDescriptors, valueRangeManager, 10);
+        } else {
+            return new OriginalFilteringValueRangeIterator<>(this::selectReplayedEntity, childEntitySelector.iterator(),
+                    basicVariableDescriptors, valueRangeManager);
+        }
+    }
+
+    @Override
+    public ListIterator<Object> listIterator() {
+        throw new UnsupportedOperationException();
```

Comment:
What is this being used for?
If we can afford not to implement it, why is it even there?

Reply:
Yeah, we won't spend time on this now, but in the future, I think the list iterator can just be replaced by a plain iterator.